File: formacOS/Sources/AnyClipCore/AuthGate.swift
import Foundation

/// Per-IP cooldown after repeated handshake failures. After maxFails failed
/// handshakes from the same IP, that IP is blocked for cooldown seconds.
/// A successful handshake clears the counter. Stale entries are swept lazily.
/// Port of anyclip.AuthGate; the caller (PeerLink actor) provides isolation.
public struct AuthGate: Sendable {
    public static let maxFails = 5
    public static let cooldown: Double = 60.0

    private struct Entry: Sendable {
        var count: Int
        var last: Double
    }

    private var fails: [String: Entry] = [:]
    private let now: @Sendable () -> Double

    public init(now: @escaping @Sendable () -> Double = { Date().timeIntervalSince1970 }) {
        self.now = now
    }

    public func isBlocked(_ ip: String) -> Bool {
        let t = now()
        guard let entry = fails[ip] else { return false }
        // Entry is stale (past cooldown window) → treat as not blocked.
        if t - entry.last >= Self.cooldown { return false }
        return entry.count >= Self.maxFails
    }

    public mutating func recordFail(_ ip: String) {
        let count = fails[ip]?.count ?? 0
        fails[ip] = Entry(count: count + 1, last: now())
        sweep()
    }

    public mutating func recordOK(_ ip: String) {
        fails[ip] = nil
    }

    private mutating func sweep() {
        let t = now()
        fails = fails.filter { t - $0.value.last < Self.cooldown }
    }
}

File: forwindows/src/AnyClipCore/AuthGate.cs
namespace AnyClip.Core;

/// Per-IP cooldown after repeated handshake failures (5 fails → 60 s).
/// Port of anyclip.AuthGate with the Swift-port fix: RecordFail sweeps
/// expired entries BEFORE reading the old count, so a stale count never
/// carries into a new window. Caller provides synchronization.
public sealed class AuthGate(Func<double>? now = null)
{
    public const int MaxFails = 5;
    public const double CooldownSeconds = 60.0;

    private readonly Func<double> _now =
        now ?? (() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0);
    private readonly Dictionary<string, (int Count, double Last)> _fails = new();

    public bool IsBlocked(string ip)
    {
        if (!_fails.TryGetValue(ip, out var e)) return false;
        if (_now() - e.Last >= CooldownSeconds) return false; // expired never blocks
        return e.Count >= MaxFails;
    }

    public void RecordFail(string ip)
    {
        Sweep();
        var count = _fails.TryGetValue(ip, out var e) ? e.Count : 0;
        _fails[ip] = (count + 1, _now());
    }

    public void RecordOk(string ip) => _fails.Remove(ip);

    private void Sweep()
    {
        var t = _now();
        foreach (var ip in _fails.Where(kv => t - kv.Value.Last >= CooldownSeconds)
                                 .Select(kv => kv.Key).ToList())
            _fails.Remove(ip);
    }
}
